src/frontend/org/voltdb/importclient/kafka/KafkaStreamImporter.java

```
@@ -653,115 +648,116 @@ public void run() {
+
+                        String line = new String(payload.array(),payload.arrayOffset(),payload.limit(),"UTF-8");
+                        CSVInvocation invocation = new CSVInvocation(m_procedure, line);
+                        TopicPartitionInvocationCallback cb = new TopicPartitionInvocationCallback(currentOffset, messageAndOffset.nextOffset(), m_topicAndPartition, cbcnt);
+                        m_pendingOffsets.add(currentOffset);
+                        if (!callProcedure(cb, invocation)) {
+                            if (isDebugEnabled()) {
+                                debug("Failed to process Invocation possibly bad data: " + line);
                             }
-                            submitCount++;
-                            m_currentOffset.set(messageAndOffset.nextOffset());
-                            if (m_shutdown) {
-                                break;
+                            synchronized(m_seenOffset) {
+                                //Make this failed offset known to seen offsets so committer can push ahead.
+                                m_seenOffset.add(messageAndOffset.nextOffset());
                             }
+                            m_pendingOffsets.remove(currentOffset);
+                        }
+                        submitCount++;
+                        m_currentOffset.set(messageAndOffset.nextOffset());
+                        if (m_shutdown) {
+                            break;
                         }
                     }
                     if (m_shutdown) {
                         break;
                     }
 
-                    if (currentFetchCount == 0) {
+                    //wait to fetch more if we are in backpressure or read nothing last time.
+                    if (currentFetchCount == 0 || m_hasBackPressure) {
                         try {
-                            Thread.sleep(1000);
+                            Thread.sleep(m_backpressureSleepMs);
```

should we really swallow this exception may we should rethrow it as sensible extension to `RuntimeException`